projects/charger/inc/charger_fsm.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+// FSM for the charger
+//
+// Requires
+// - notify
+// - charger_pin
+// - charger_controller
+// To be initialized.
+//
+// Charger design:
+//
+// Charger Disconnected:
+// - Do nothing
+//
+// Charger Connected:
+// - Notify of connection periodically
+//
+// Charger Charging:
+// - Notify of connection periodically
+// - Charge while command is not expired and periodically publish data
+// - Exits immediately in case of fault
+//
+// Transition Table:
+// | From \ To    | Disconnected | Connected | Charging |
+// | Disconnected |              |     Y     |     N    |
+// | Connected    |       Y      |           |   G(Y)   |
+// | Charging     |       Y      |           |          |
+//
+// Y - Allowed
+// N - Not Allowed
+// G(Y) - Guarded but Allowed
+
+#include <stdbool.h>
+
+#include "charger_can.h"
+#include "event_queue.h"
+
+// Initializes the FSM |charger_can_status| is used to guard against faults.
+void charger_fsm_init(ChargerCanStatus *charger_can_status);
+
+// Attempts the transition based on the event in |e|.
+bool charger_fsm_process_event(const Event *e);
```

Comment:
Done. I don't really like that design pattern. I would rather explicitly declare which FSM are executing with this kind of function call but since this seems to be the paradigm others are using I guess I'll go with it.1